kong/clustering/services/sync/init.lua

```diff
@@ -0,0 +1,69 @@
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local events = require("kong.clustering.events")
+local strategy = require("kong.clustering.services.sync.strategies.postgres")
+local rpc = require("kong.clustering.services.sync.rpc")
+
+
+-- TODO: what is the proper value?
+local FIRST_SYNC_DELAY = 0.5  -- seconds
+local EACH_SYNC_DELAY  = 30   -- seconds
+
+
+function _M.new(db, is_cp)
+  local strategy = strategy.new(db)
+
+  local self = {
+    db = db,
+    strategy = strategy,
+    rpc = rpc.new(strategy),
+    is_cp = is_cp,
+  }
+
+  -- only cp needs hooks
+  if is_cp then
+    self.hooks = require("kong.clustering.services.sync.hooks").new(strategy)
+  end
+
+  return setmetatable(self, _MT)
+end
+
+
+function _M:init(manager)
+  if self.hooks then
+    self.hooks:register_dao_hooks()
+  end
+  self.rpc:init(manager, self.is_cp)
+end
+
+
+function _M:init_worker()
+  -- is CP, enable clustering broadcasts
+  if self.is_cp then
+    events.init()
+
+    self.strategy:init_worker()
+    return
+  end
+
+  -- is DP, sync only in worker 0
+  if ngx.worker.id() ~= 0 then
+    return
+  end
+
+  -- sync to CP ASAP
+  assert(self.rpc:sync_once(FIRST_SYNC_DELAY))
+
+  assert(ngx.timer.every(EACH_SYNC_DELAY, function(premature)
+    if premature then
+      return
+    end
+
+    assert(self.rpc:sync_once())
+  end))
```

Comment:
This logic starts two timers every 30 seconds. For further optimization, we could only start one timer. We can place a TODO here. Since we have encountered some resource leak problem with timers, I think it would be better to create as few timers as possible.

The current logic looks like the following:

```
timer.every (30s, function
      sync_once()
         '-> timer.at(0, sync_timer_function )
end

function sync_once()
     timer.at(0, function_to_sync )
end
```

We could optimize it to the following logic

```
timer.every(30s, function_to_sync )
```

Reply:
hi @chronolaw i have an idea to resolve this, we could introduce an API like `self.rpc:sync_every()`, then the logic looks like this:

```
local function _sync_once_function(premature)
...
end

function _M:sync_once(delay)
  local hdl, err = ngx.timer.at(delay or 0, _sync_once_function)
end

function _M:sync_every(delay)
  local hdl, err = ngx.timer.every(delay or 0, _sync_once_function)
end

```

I'll open a PR for this, could you help review

Reply:
try to fix it in https://github.com/Kong/kong/pull/13732

Reply:
Merged, thanks.